cloudinit/cmd/devel/hotplug_hook.py

```diff
@@ -200,29 +221,30 @@ def handle_args(name, args):
     log.setupLogging(hotplug_init.cfg)
     if 'reporting' in hotplug_init.cfg:
         reporting.update_configuration(hotplug_init.cfg.get('reporting'))
-
     # Logging isn't going to be setup until now
     LOG.debug(
-        '%s called with the following arguments: {udevaction: %s, '
-        'subsystem: %s, devpath: %s}',
-        name, args.udevaction, args.subsystem, args.devpath
-    )
-    LOG.debug(
-        '%s called with the following arguments:\n'
-        'udevaction: %s\n'
-        'subsystem: %s\n'
-        'devpath: %s',
-        name, args.udevaction, args.subsystem, args.devpath
+        '%s called with the following arguments: {'
+        'hotplug_action: %s, subsystem: %s, udevaction: %s, devpath: %s}',
+        name,
+        args.hotplug_action,
+        args.subsystem,
+        args.udevaction if 'udevaction' in args else None,
+        args.devpath if 'devpath' in args else None,
     )
 
     with hotplug_reporter:
         try:
-            handle_hotplug(
-                hotplug_init=hotplug_init,
-                devpath=args.devpath,
-                subsystem=args.subsystem,
-                udevaction=args.udevaction,
-            )
+            if args.hotplug_action == 'query':
+                hotplug_init.fetch(existing="trust")
```

Comment:
If we run this before the datasource is detected, we trace here:
```bash
cloudinit.sources.DataSourceNotFoundException: Did not find any data source, searched classes: (DataSourceNone)
```
We might want to catch that case and at least report "unknown" or "no-datasource" etc.
```suggestion
                try:
                    hotplug_init.fetch(existing="trust")
                except cloudinit.sources.DataSourceNotFoundException:
                     print("Unable to determine hotplug state. No datasource detected")
                     sys.exit(1)              
```

Also is it worth just baking this hotplug_init.fetch(existing="trust") into is_enabled() function because I see this same potential problem in handle_hotplug too?

I realize generally this cmdline shouldn't normally get run until after the datasource detection has occurred, just a race there until that discovery happens.

Reply:
Our udev hook should disable any hotplug commands until cloud-init has finished its initial run. How did you get this exception to happen? I don't necessarily disagree with your suggestion, just trying to understand.